prow/config/config_test.go

```diff
@@ -87,33 +87,75 @@ func volumesAndMountsMatch(mounts []v1.VolumeMount, volumes []v1.Volume) bool {
+	malformed := map[string]sets.String{}
+	for _, container := range spec.InitContainers {
+		malformed[container.Name] = missingVolumesForContainer(container.VolumeMounts, spec.Volumes)
+	}
+	for _, container := range spec.Containers {
+		malformed[container.Name] = missingVolumesForContainer(container.VolumeMounts, spec.Volumes)
+	}
+	return malformed
+}
+
+func missingMountsForSpec(spec *v1.PodSpec) sets.String {
+	mountNames := sets.NewString()
+	volumeNames := sets.NewString()
 	for _, container := range spec.Containers {
-		if !volumesAndMountsMatch(container.VolumeMounts, spec.Volumes) {
-			return false
+		for _, m := range container.VolumeMounts {
+			mountNames.Insert(m.Name)
+		}
+	}
+	for _, container := range spec.InitContainers {
+		for _, m := range container.VolumeMounts {
+			mountNames.Insert(m.Name)
 		}
 	}
-	return true
+	for _, v := range spec.Volumes {
+		volumeNames.Insert(v.Name)
+	}
+	return volumeNames.Difference(mountNames)
 }
 
 // verify that all volume mounts reference volumes that exist
 func TestMountsHaveVolumes(t *testing.T) {
 	for _, job := range c.AllPresubmits(nil) {
-		if job.Spec != nil && !specHasMatchingVolumesAndMounts(job.Spec) {
-			t.Errorf("volumes and mounts do not match for: %v", job.Name)
+		if job.Spec != nil {
+			for container, missingVolumes := range missingVolumesForSpec(job.Spec) {
+				if len(missingVolumes) > 0 {
+					t.Errorf("job %s in container %s has mounts that are missing volumes: %v", job.Name, container, missingVolumes.List())
+				}
+			}
+			if missingMounts := missingMountsForSpec(job.Spec); len(missingMounts) > 0 {
+				t.Errorf("job %s has volumes that are not mounted: %v", job.Name, missingMounts.List())
+			}
 		}
 	}
 	for _, job := range c.AllPostsubmits(nil) {
-		if job.Spec != nil && !specHasMatchingVolumesAndMounts(job.Spec) {
-			t.Errorf("volumes and mounts do not match for: %v", job.Name)
+		if job.Spec != nil {
+			for container, missingVolumes := range missingVolumesForSpec(job.Spec) {
+				if len(missingVolumes) > 0 {
+					t.Errorf("job %s in container %s has mounts that are missing volumes: %v", job.Name, container, missingVolumes.List())
+				}
+			}
+			if missingMounts := missingMountsForSpec(job.Spec); len(missingMounts) > 0 {
+				t.Errorf("job %s has volumes that are not mounted: %v", job.Name, missingMounts.List())
+			}
 		}
 	}
 	for _, job := range c.AllPeriodics() {
-		if job.Spec != nil && !specHasMatchingVolumesAndMounts(job.Spec) {
-			t.Errorf("volumes and mounts do not match for: %v", job.Name)
+		if job.Spec != nil {
+			for container, missingVolumes := range missingVolumesForSpec(job.Spec) {
+				if len(missingVolumes) > 0 {
+					t.Errorf("job %s in container %s has mounts that are missing volumes: %v", job.Name, container, missingVolumes.List())
+				}
+			}
+			if missingMounts := missingMountsForSpec(job.Spec); len(missingMounts) > 0 {
+				t.Errorf("job %s has volumes that are not mounted: %v", job.Name, missingMounts.List())
+			}
```

Comment:
This code appears to be common for all job types. Can you define it once in a function that takes the job name, spec, and a `testing.T` instead of copying it?

Reply:
Just when I thought I could be lazy and get away with it :)